### PR TITLE
Store remote dataset credentials separately

### LINKED ISSUE #6612: Store and manage credentials separate from datasets
## Detailed Description
Credentials for remote datasets should be stored and managed separately from datasets. The credentials should be created by dataset managers or admins. Credentials should never be sent back to the client. Credentials can then be attached to layers (or mags).
Not sure if we should allow to delete credentials that are referenced by a dataset. Pro: It is very easy to revoke access. Con: It makes datasets unusable and maybe hard to find out why.

### PR DESCRIPTION
**Pull request by frcroth**
### URL of deployed dev instance (used for testing):
- https://storecredentialsseparately.webknossos.xyz/

### Steps to test:
- Add a dataset with credentials (S3 or HTTP Basic Auth) via exploration
- Test that the dataset can be viewed
- Test that the following other scenarios are still possible
	- Viewing a dataset without credentials
	- Viewing  a dataset with legacy credentials in the json
- Test the POST route to create a credential (This route is currently not used)

### TODO
- [x] Do not put passwords in log (this is currently the case because the body of the rpc request is printed)
- [x] cache bucket providers – they are no longer vals in layer, but now defs, so there might be new ones created all the time (I can do this –@fm3)

### Issues:
- fixes #6612

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [x] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [x] Needs datastore update
- [x] Ready for review

### CODE REVIEW

**Comment by frcroth** [comment]
Waiting for #6679 to fix the s3 file system.

**Review by fm3 — CHANGES_REQUESTED** [review]
Cool stuff, works really well :) I added a few refactoring comments, but nothing major.

File: app/models/binary/credential/Credential.scala [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+import javax.inject.Inject
+import scala.concurrent.ExecutionContext
+
+// Generic credential as it appears in the database
+case class Credential(_id: ObjectId,
+                      credentialType: CredentialType.Value,
+                      name: String,
+                      identifier: Option[String],
+                      secret: Option[String],
+                      scope: Option[String],
```
Comment (fm3) [inline app/models/binary/credential/Credential.scala:20]: scope is no longer included, right?

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/storage/FileSystemCredentials.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.scalableminds.webknossos.datastore.storage
+
+import play.api.libs.json.{Json, OFormat}
+
+sealed trait AnyCredential
+
+object AnyCredential {
+  implicit val jsonFormat: OFormat[AnyCredential] = Json.format[AnyCredential]
+}
```
[the remark this answers is not in the document]
Comment (frcroth) [reply webknossos-datastore/app/com/scalableminds/webknossos/datastore/storage/FileSystemCredentials.scala:9]: Yes

```
[error] /home/felix/scm/webknossos/webknossos-datastore/app/com/scalableminds/webknossos/datastore/services/DSRemoteWebKnossosClient.scala:130:27: No Json deserializer found for type com.scalableminds.webknossos.datastore.storage.AnyCredential. Try to implement an implicit Reads or Format for this type.
[error]       .getWithJsonResponse[AnyCredential]
[error]                           ^
[error] one error found
[error] (webknossosDatastore / Compile / compileIncremental) Compilation failed
```

**Review by fm3 — APPROVED** [review]
LGTM :) Please have a look at the merge conflicts, including evolution numbering.

Note that refresh-schema will delete the stored credentials for local datasets, so streaming test datasets become unusable then. I’m not sure how to manage this for our local development. May not be a serious problem. The old format still works, right? So maybe we could manually edit the json for a local dataset to include the credentials again, to keep it locally for testing across schema refreshs.

**Pull request merged**